ipns/record.go

```diff
@@ -3,124 +3,466 @@ package ipns
+//
+//   - -1 if a is older than b
+//   - 0 if a and b cannot be ordered (this doesn't mean that they are equal)
+//   - +1 if a is newer than b
+//
+// This function does not validate the records. The caller is responsible for
+// ensuring that the Records are valid by using [Validate].
+func compare(a, b *Record) (int, error) {
+	aHasV2Sig := a.pb.GetSignatureV2() != nil
+	bHasV2Sig := b.pb.GetSignatureV2() != nil
+
+	// Having a newer signature version is better than an older signature version
+	if aHasV2Sig && !bHasV2Sig {
+		return 1, nil
+	} else if !aHasV2Sig && bHasV2Sig {
+		return -1, nil
 	}
 
-	var i int
-	for j := 1; j < len(recs); j++ {
-		cmp, err := Compare(recs[i], recs[j])
+	as, err := a.Sequence()
+	if err != nil {
+		return 0, err
+	}
+
+	bs, err := b.Sequence()
+	if err != nil {
+		return 0, err
+	}
+
+	if as > bs {
+		return 1, nil
+	} else if as < bs {
+		return -1, nil
+	}
+
+	at, err := a.Validity()
+	if err != nil {
+		return 0, err
+	}
+
+	bt, err := b.Validity()
+	if err != nil {
+		return 0, err
+	}
+
+	if at.After(bt) {
+		return 1, nil
+	} else if bt.After(at) {
+		return -1, nil
+	}
+
+	return 0, nil
+}
+
+// EmbedPublicKey embeds the given public key in the given [Record]. While not
+// strictly required, some notes (e.g., DHT servers), may reject IPNS Records
+// that do not embed their public keys as they may not be able to validate them
+// efficiently.
+func EmbedPublicKey(r *Record, pk ic.PubKey) error {
+	// Try extracting the public key from the ID. If we can, do not embed it.
+	pid, err := peer.IDFromPublicKey(pk)
+	if err != nil {
+		return err
+	}
+	if _, err := pid.ExtractPublicKey(); err != peer.ErrNoPublicKey {
+		// Either a *real* error or nil.
+		return err
+	}
+
+	// We failed to extract the public key from the peer ID, embed it.
+	pkBytes, err := ic.MarshalPublicKey(pk)
+	if err != nil {
+		return err
+	}
+
+	r.pb.PubKey = pkBytes
+	return nil
+}
+
+// ExtractPublicKey extracts a [crypto.PubKey] matching the given [peer.ID] from
+// the IPNS Record, if possible.
+func ExtractPublicKey(r *Record, pid peer.ID) (ic.PubKey, error) {
+	if pk, err := r.PubKey(); err == nil {
+		expPid, err := peer.IDFromPublicKey(pk)
 		if err != nil {
-			return -1, err
-		}
-		if cmp == 0 {
-			cmp = bytes.Compare(vals[i], vals[j])
+			return nil, multierr.Combine(ErrInvalidPublicKey, err)
 		}
-		if cmp < 0 {
-			i = j
+
+		if pid != expPid {
+			return nil, ErrPublicKeyMismatch
 		}
+
+		return pk, nil
+	} else if !errors.Is(err, ErrPublicKeyNotFound) {
+		return nil, multierr.Combine(ErrInvalidPublicKey, err)
+	} else {
+		return pid.ExtractPublicKey()
+	}
+}
+
+// RoutingKey returns the IPNS routing key for the given [peer.ID]. Note
+// that the intended use of this function is for routing only. The output of this
+// function is not human readable. For a human readable version, see [Key].
+func RoutingKey(pid peer.ID) string {
+	return "/ipns/" + string(pid)
+}
```

Comment:
Could we make this return `byte[]`  and not `string`, just to make it extra clear this is binary and people who want string should use the `Key` one?

```suggestion
// RoutingKey returns the binary IPNS routing key for the given [peer.ID]. Note
// that the intended use of this function is for routing only. The output of this
// function is binary, not human readable. For a string, human readable version, see [Key].
func RoutingKey(pid peer.ID) byte[] {
    var buffer bytes.Buffer
	buffer.WriteString("/ipns/")
	buffer.Write(pid)
	return buffer.Bytes()
}
```

Reply:
@lidel I would be all up for this idea, but the libp2p `record.Validator` interface requires a `key string` 😃 If we do return `[]byte`, then the caller would always have to transform it to a `string` 🙃

Reply:
I know, I think that is a feature: it highlights the problem to external users, and does not try to hide it inside our library.
If we ever refactor record  Validator interface, it will make it easier to switch to `byte[]` there.

An alternative would be to add a specific helper, like `BinaryRoutingKeyAsString` to make it extra clear and not hide conversion.

Either works, just want to help users understand why there is `/ipfs/Ò̴̺̱̝̹̀̆̃͑ͧ̕͞m̴̡͙̫̹͕̲̱͔̗̹̹̺̺̤̼̺͉͚̙̌ͮ̀̌̓͋̅̏̈́̓̅̍̾̓̚g̛͙̬̬̥̜̫̮͖͉̱̗̟͎̘̓ͪ͊̿̽ͥͮ͗̏̓͐͂ͬ͆ͨͨ͂͝ ̡͖̠͓̭͖̯̲̬͈̗̭̈̆͆̈̒̇̓͋̓̉ͩͭͥ̒̓ͮ͝ͅw̺͓͓̼̬̯̲̪̙̰̗͙̞̳͕͇̅̑ͨ̍ͮ̈́̃͋͂̓ͬ͒͜͜͢͡h̴͖̤̦̱̰̞̲̱͙̭̗̫͚̗͂̏ͧ͌̃ͤ͋̅ͭ̎̌̆͑̔͢ͅa̷̛̰̻͉̟̻̦̟̱͉̘̘̭̩̗͇͙̪͑̐̑ͨ̾ͪ̾̊̒̕͝ͅt͔̰͓̬̼̖̟͙̑̏̇͐̐͆͋ͯ͠ ̷̛͚̣̤̝̜͌͌ͪͣ͋ͤ̃ͮͣͫ̿̇͝i̷̡̜͕̱͍͉̽ͥ̀̾̀̀ş̶̼͕̖͔͎̦͗͋̍ͯ̔ͨͪ̓͑̐͑̈́́ ̸̱̪͈͓̹̏ͩͪ͊͑ͭ͝ţ̷͗̐͋̍ͮ̇̓͑͋̚͏̖̹̹̬̦̻̭͓̯͔̜̼̯̹̯ḩ̴̘̯̠̯̋̎̑̋́ͦ͝ͅi̴̧̾ͣ̀̌̋̆̈́͏̮͇͓͇͕̪̠͇̻͇͖̝̳̞͙s̴̷̐ͥ̌͆ͣͩ̓ͬ̽ͫ҉͏̫̦͈̱̖̭͎̰̺͙̥̩̩͉͙̞̳ͅ` in their logs somehow, especially when they look at Go code when trying to reimplement things in other languages.  :see_no_evil:

Reply:
I'll change it then, hopefully the validator will be updated at some point 🙃